### PR TITLE
[java] 1862 - Message Digest should not be used as class field

### PR DESCRIPTION
**Pull request by AnthonyKot**
Before submitting a PR, please check that:
 - [x] The PR is submitted against `master`. The PMD team will merge back to support branches as needed.
 - [x] `./mvnw clean verify` passes. This will [build](https://github.com/pmd/pmd/blob/master/BUILDING.md) and test PMD, execute PMD and checkstyle rules. [Check this for more info](https://github.com/pmd/pmd/blob/master/CONTRIBUTING.md#code-style)

**PR Description:**
New rule which checks there is no fields of MessageDigest type is added.
Based on this discussion:
https://github.com/pmd/pmd/pull/1963
and this issue:
https://github.com/pmd/pmd/issues/1862

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/67e8373c64424c60be5208e7777ed7c0/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/resources/category/java/bestpractices.xml [hunk 1]
```diff
@@ -1710,4 +1710,58 @@ public class Example {
...
+                <value>
+                    <![CDATA[
+//FieldDeclaration/Type/ReferenceType/ClassOrInterfaceType[@Image = 'MessageDigest']
+]]>
+                </value>
+            </property>
+        </properties>
+        <example>
+            <![CDATA[
+public class Foo {
```
Comment (adangel) [inline pmd-java/src/main/resources/category/java/bestpractices.xml]: I think, the example is too big and misleading (appart from the problem, that is doesn't compile).
It also doesn't show, how to do it correctly.

```java
import java.security.MessageDigest;
public class AvoidMessageDigestFieldExample {
    private final MessageDigest sharedMd;
    public AvoidMessageDigestFieldExample() throws Exception {
        sharedMd = MessageDigest.getInstance("SHA-256");
    }
    public byte[] calculateHashShared(byte[] data) {
        // sharing a MessageDigest like this without synchronizing access
        // might lead to wrong results
        sharedMd.reset();
        sharedMd.update(data);
        return sharedMd.digest();
    }

    // better
    public byte[] calculateHash(byte[] data) throws Exception {
        MessageDigest md = MessageDigest.getInstance("SHA-256");
        md.update(data);
        return md.digest();
    }
}
```

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidMessageDigestField.xml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<test-data
+    xmlns="http://pmd.sourceforge.net/rule-tests"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://pmd.sourceforge.net/rule-tests http://pmd.sourceforge.net/rule-tests_1_0_0.xsd">
+    <test-code>
+        <description><![CDATA[
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidMessageDigestField.xml]: The CDATA-section is here not necessary. Just enter the description directly.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidMessageDigestField.xml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        <code><![CDATA[
+import java.security.MessageDigest;
+
+public class Foo {
+    private MessageDigest md;
+    private static MessageDigest staticMd;
+}
+     ]]></code>
+    </test-code>
+</test-data>
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidMessageDigestField.xml]: We definitively need an example, that verifies, that we don't flag *valid* uses of MessageDigest. We don't want to introduce false-positive...

File: pmd-java/src/main/resources/category/java/bestpractices.xml [hunk 4]
```diff
@@ -1710,4 +1710,58 @@ public class Example {
         </example>
     </rule>
 
+    <rule name="AvoidMessageDigestField"
```
[the remark this answers is not in the document]
Comment (AnthonyKot) [reply pmd-java/src/main/resources/category/java/bestpractices.xml]: Done

File: pmd-java/src/main/resources/category/java/bestpractices.xml [hunk 5]
```diff
@@ -1710,4 +1710,58 @@ public class Example {
...
+          language="java"
+          minimumLanguageVersion="1.5"
+          since="6.18.0"
+          message="You shouldn't declare field of MessageDigest type, because unsynchronized access could cause problems"
+          class="net.sourceforge.pmd.lang.rule.XPathRule"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_java_bestpractices.html#avoidmessagedigestfield">
+        <description>
+            Since Java5 brought a new implementation of the Map designed for multi-threaded access, you can
```
[the remark this answers is not in the document]
Comment (AnthonyKot) [reply pmd-java/src/main/resources/category/java/bestpractices.xml]: Forgot to replace in new commit. Fixed.

File: pmd-java/src/main/resources/category/java/bestpractices.xml [hunk 6]
```diff
@@ -1710,4 +1710,58 @@ public class Example {
         </example>
     </rule>
 
+    <rule name="AvoidMessageDigestField"
+          language="java"
+          minimumLanguageVersion="1.5"
```
[the remark this answers is not in the document]
Comment (AnthonyKot) [reply pmd-java/src/main/resources/category/java/bestpractices.xml]: You'r right, at least from Java 3.

File: pmd-java/src/main/resources/category/java/bestpractices.xml [hunk 7]
```diff
@@ -1710,4 +1710,58 @@ public class Example {
...
+        <description>
+            Since Java5 brought a new implementation of the Map designed for multi-threaded access, you can
+            perform efficient map reads without blocking other threads.
+        </description>
+        <priority>3</priority>
+        <properties>
+            <property name="xpath">
+                <value>
+                    <![CDATA[
+//FieldDeclaration/Type/ReferenceType/ClassOrInterfaceType[@Image = 'MessageDigest']
```
[the remark this answers is not in the document]
Comment (AnthonyKot) [reply pmd-java/src/main/resources/category/java/bestpractices.xml]: Has sense

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidMessageDigestField.xml (hunk 3 above)
Comment (AnthonyKot) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidMessageDigestField.xml]: Added

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidMessageDigestField.xml (hunk 2 above)
Comment (AnthonyKot) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidMessageDigestField.xml]: Ok

File: pmd-java/src/main/resources/category/java/bestpractices.xml (hunk 1 above)
Comment (AnthonyKot) [reply pmd-java/src/main/resources/category/java/bestpractices.xml]: I put this example. Thank you

**Pull request merged**